mivideo/mivideo_extract.py

```diff
@@ -67,62 +88,248 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+
+        :return: a dictionary with ValidDataSourceName and last run timestamp
+        """
+
+        self._kalturaInit()
+
+        KALTURA_MAX_MATCHES_ERROR: str = 'QUERY_EXCEEDED_MAX_MATCHES_ALLOWED'
+        tableName: str = 'mivideo_media_created'
+
+        localLogger = logging.getLogger(f'{logger.name}.mediaCreation')
+        localLogger.info('Starting procedure...')
+
+        kClient: KalturaRequestConfiguration = KalturaClient(KalturaConfiguration())
+        kClient.setKs(  # pylint: disable=no-member
+            KalturaSessionService(kClient).start(
+                self.kUserSecret, type=KalturaSessionType.ADMIN, partnerId=self.kPartnerId))
+        kMedia = KalturaMediaService(kClient)
+
+        lastTime: Union[datetime, None] = (
+            self._readTableLastTime(tableName, 'created_at'))
+
+        if (lastTime):
+            localLogger.info(
+                f'Last time found in table: "{lastTime.isoformat()}"')
+        else:
+            lastTime = datetime.fromisoformat(self.defaultLastTimestamp)
+            localLogger.info('Last time not found in table; '
+                             f'using default time: "{lastTime.isoformat()}"')
+
+        createdAtTimestamp: float = lastTime.timestamp()
+
+        kFilter = KalturaMediaEntryFilter()
+        kFilter.createdAtGreaterThanOrEqual = createdAtTimestamp
+        kFilter.categoriesFullNameIn = self.categoriesFullNameIn
+        kFilter.orderBy = KalturaMediaEntryOrderBy.CREATED_AT_ASC
+
+        kPager = KalturaFilterPager()
+        kPager.pageSize = 500  # 500 is maximum
+        kPager.pageIndex = 1
+
+        results: Sequence[KalturaMediaEntry] = None
+        lastCreatedAtTimestamp: Union[float, int] = createdAtTimestamp
+        lastId: Union[str, None] = None
+        numberResults: int = 0
+        queryPageNumber: int = kPager.pageIndex  # for logging purposes
+        totalNumberResults: int = numberResults  # for logging purposes
+        endOfResults = False
+
+        while not endOfResults:
+            try:
+                results = kMedia.list(kFilter, kPager).objects
+            except KalturaException as kException:
+                if (KALTURA_MAX_MATCHES_ERROR in kException.args):
+                    # set new filter timestamp, reset pager to page 1, then continue
+                    kFilter.createdAtGreaterThanOrEqual = lastCreatedAtTimestamp
+                    localLogger.debug(
+                        f'New filter timestamp: ({kFilter.createdAtGreaterThanOrEqual})')
+
+                    # to avoid dupes, also filter out the last ID returned by previous query
+                    # because Kaltura compares createdAt greater than *or equal* to timestamp
+                    kFilter.idNotIn = lastId
+                    kPager.pageIndex = 1
+                    continue
+
+                localLogger.info(f'Other Kaltura API error: "{kException}"')
+                break
```

Comment:
I don't quite understand the KALTURA_MAX_MATCHES_ERROR (what does that mean?), but have you not had any issues with other connection-related exceptions? Seems like we might need some re-try behavior?

Reply:
Kaltura allows exactly 10,000 results from any API query.  If you attempt to read past that, they throw a `KalturaException` with a message that contains the string `QUERY_EXCEEDED_MAX_MATCHES_ALLOWED`.  Unfortunately, they don't throw a specific exception for this condition.  They use that general one with a string.

The way to handle that error is to change the filter and query again.  In our case, since we're querying for things newer than a specific time and they're sorted increasingly, we can make a new query by using the last time received in the API's results.  When we use a new time in subsequent API queries, we also must exclude the last item received by ID, because their filter is for `createdAtGreaterThanOrEqual`.  They don't have a strictly equal filter.  🙁 

So, yes, this **_is_** retry behavior.  I've not seen other connection problems, but I've also not tried to induce them.  We may need handling for that.  I'd suggest that should be a new issue.

Reply:
I think we do need handling for other exceptions, but I am okay with you creating a separate issue, so long as this is handled in the near future and not postponed indefinitely. If what we're going through with Canvas right now can be extrapolated, we cannot assume that API endpoints will always perform perfectly.

Reply:
Opened issue #148.